src/server/pkg/tarutil/util.go

```diff
@@ -0,0 +1,106 @@
+package tarutil
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/tar"
+)
+
+type File interface {
+	Header() (*tar.Header, error)
+	Content(io.Writer) error
+}
+
+type file struct {
+	hdr  *tar.Header
+	data []byte
+}
+
+func NewFile(name string, data []byte) File {
+	return newFile(NewHeader(name, int64(len(data))), data)
+}
+
+func newFile(hdr *tar.Header, data []byte) File {
+	return &file{
+		hdr:  hdr,
+		data: data,
+	}
+}
+
+func NewHeader(name string, size int64) *tar.Header {
+	return &tar.Header{
+		Name: name,
+		Size: size,
+	}
+}
+
+func (f *file) Header() (*tar.Header, error) {
+	return f.hdr, nil
+}
+
+func (f *file) Content(w io.Writer) error {
+	_, err := w.Write(f.data)
+	return err
+}
+
+func WriteFile(tw *tar.Writer, file File) error {
+	hdr, err := file.Header()
+	if err != nil {
+		return err
+	}
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+	if err := file.Content(tw); err != nil {
+		return err
+	}
+	return tw.Flush()
+}
+
+func WithWriter(w io.Writer, cb func(*tar.Writer) error) (retErr error) {
+	tw := tar.NewWriter(w)
+	defer func() {
+		if retErr == nil {
+			retErr = tw.Close()
+		}
+	}()
+	return cb(tw)
+}
+
+func Iterate(r io.Reader, cb func(File) error) error {
+	tr := tar.NewReader(r)
+	for {
+		hdr, err := tr.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
```

Comment:
I assume this is like a root cause thing? Idk if it's appropriate for a stop iteration sentinel like `io.EOF`.

Reply:
Put another way: the protocol is "io.EOF signals the end of the tar stream". Not "something happening which has a root, intermediate, or proximate cause of io.EOF" signals the end of the tar stream.

Reply:
There is another error type called `io.ErrUnexpectedEOF` that is intended for actual errors, and it will be emitted by the tar reader when unexpected `io.EOF`s occur. `io.EOF` is a pretty standard sentinel for the successful end of a stream / file.

Reply:
I agree that it is a standard sentinel.  It would be nonstandard to signal the end of iteration with a wrapped error. I don't know of anything that does that. And it seems weird for this code to expect or at least handle that case.

Reply:
Ahh, I see. Yeah, I'm not too sure. I think I would prefer to play it safe for the time being.